### ISSUE TITLE
[RFC] A more structured API for parallel I/O

### ISSUE THREAD

**Issue by zah** [status-im/nim-asyncdispatch2#2]
#### 1. Allow `await` to be given multiple arguments or a tuple 

``` nim
proc httpRequest(url: string): Future[HttpResult]
proc jsonRpcCall(url: string): JsonNode

proc foo {.async.} =
  var keyword = "Status"
  var (googlePage, jsonData) = await(httpRequest(&"http://google.com/?q={keyword}"),
                                     jsonRpcCall("localhost/myApi"))
  
  echo "HTTP RESPONSE ", googlePage.status, "\n", googlePage.body
  echo "JSON RESPONSE\n", jsonData{"result"}
```

This form of `await` just performs the I/O operations in parallel returning a tuple of the final results. It is similar to using `var r1 = request(); var r2 = request(); await all(r1, r2)` in the current design.

For convenience `await (foo, bar)` is considered the same as `await(foo,bar)`.

#### 2. Introduce a new `select` API:

`select` is a new API that is given a number of I/O operations that should be started in parallel. The key difference from `await` is that the handlers installed for each operation will be executed as soon as any of the results are ready and control flow keywords such as `return` and `break` can be used to cancel some of the outstanding operations:

``` nim
proc foo {.async.} =
  var keyword = "Status"
  var timedOut = false

  select:
    httpRequest(&"http://google.com/?q={keyword}") as response:
      # executes this as soon as it's ready
      echo "HTTP RESPONSE ", googlePage.status, "\n", googlePage.body
 
    jsonRpcCall("localhost/myApi") as jsonData:
      echo jsonData{"result"}
      return # returns from the current proc; skips the other handlers

    *timeout(100):
      # `timeout` is the same as `sleepAsync`
      timedOut = true
      break # continues after the select; skips the other handlers

  echo "async ops ", if timedOut: "timed out" else: "finished"
```

The execution after the `select` block continues when all of the handlers have been executed, although there must be a way to mark some of them as optional (here, I've used `*` for this). 

#### 3. Introduce a new `safeasync` pragma

The `safeasync` pragma is responsible for inserting the `await` keyword in automatic way. It also has the role of the current `multisync` pragma in the sense that it allows you to compile the same code for both sync and async usage:

``` nim
proc foo: bool {.safeasync.} =
  var keyword = "Status"
  # Notice how I don't need to use await anymore
  var (googlePage, jsonData) = (httpRequest(&"http://google.com/?q={keyword}"),
                                jsonRpcCall("localhost/myApi"))

  return googlePage.status == 200 and not jsonData.hasKey("error")
```

How does this work? It inserts a call to a template called `implicitAwait` on each expression within the proc's body. `implicitAwait` is defined as an identity for all non-future types and as a regular `await` statement for all futures:

``` nim
template implicitAwait(x: auto): auto = x
template implicitAwait(x: Future): auto = await x
```

Why do I call it "safe"? Because it enforces the "structural" use of the API described above. It's not possible to make the typical programming error of not awaiting an async operation:

``` nim
proc bad(s: AsyncSocket) {.async.} =
  var status = x.send(...)
  x.close() # oops, the socket was closed prematurely
            # because the send call was not awaited.
```

Please note that the body of a `safeasync` will work in synchronous mode by executing each operation in turn. It's also possible to compile the code for implicit off-loading to a background thread pool.

### Rationale for the API

Recommended reading:
https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/

This API captures better the intention of the programmer and eliminates some avenues for errors. The "structural" higher-level abstractions enable some optimizations that are otherwise hard to achieve (such as storing the `Future` values on the stack). Manual resource management with destructors becomes completely safe and easy.

**Comment by cheatfate** [comment]
Both `point 1` and `point 2` can be replaced by primitive like this: https://docs.python.org/3/library/asyncio-task.html#asyncio.wait. 

From my experience, there no need on
```nim
var (a, b, c) = await(afut, bfut, cfut)
```
because it used very rare in real world, most of the time you filling array of tasks and waiting for any of them or for all of them.

**Comment by zah** [comment]
The dynamic interface in Python allows for working with heterogeneous result types, while in Nim you have to resort to having named `Future` variables from which the results can be extracted (losing the "structured control-flow" aspect of the design). 

The key difference in the new proposal is that the result variables have the proper type in all places and the explicit use of futures is hidden.

Otherwise, I've considered the case of requesting a sequence of homogenous operations. I think this should be a separate request throttling API.

**Comment by cheatfate** [comment]
`asyncdispatch` design allows you to accept `seq[FutureBase]` or `varargs[FutureBase]` and its enough to check readiness of `Future`.
Also extracting results from `Future` is extracting exceptions too, with your proposal its unclear how this construction must work if one of `Future` will fail, while python's `wait` allow you to specify how to react on `Exception`.

**Comment by zah** [comment]
I don't argue about the low-level details. Nothing about `asyncdispatch` should change. The goal is to remove the explicit use of futures from the high level interface, so certain optimizations are easier and the code is safer (please read the linked article).

I don't think there is an issue with exceptions. The semantics of `await` are clear - all operations must complete, else an exception will be raised. It's also possible to introduce additional syntax in the API to let you preserve some of the results as futures (e.g. `await (noraise request(), bar())`).

With `select`, again all mandatory operations must complete, otherwise an exception will be thrown. We can also allow an `except` branch for each request giving you a chance to handle the exception manually. Additional syntax will also be required for implementing progress events.

**Comment by cheatfate** [comment]
Sorry but i dont see any optimizations or safe code. Everything i see is proposal to include some more primitives to existing primitives, and add even more to resolve all possible issues.
Also i don't want to think about `sync` mode, there no reason to think about it, if people wants to be synchronous why they need asynchronous overhead.

What is proposed `await` modification? Its obfuscated version of
```python
await wait(abc, bcd, ALL_COMPLETED)
```
What is proposed `select`? Its obfuscated version of 
```python
await wait(abc, bcd, cde, ALL_COMPLETED)
```
Most of people do not understand asynchronous programming with current API, so why do you think adding more API will help it understand better?

**Comment by zah** [comment]
@cheatfate, if it's possible to design the system in a way that avoids heap allocations and reliance on the GC, it would be silly not to take advantage of it. 

The specific optimization being discussed here is placing the Future objects within the stack of the caller. At the moment this is still the heap-allocated environment of the closure iterator, but this can be attacked with further optimizations in the compiler (escape analysis for the closure objects, better support for closure/iterator monomorphism in the run-time code). In the end it may be possible to store all the required state within the single value associated with the registered I/O operation. Rust achieves this.

Also, many async procs don't actually need to be compiled as iterators:

``` nim
proc foo: Future[int] {.async.} =
  code code code  
  ...
  let v = await bar()
  ...

  return v
```

is equivalent to the sync proc:

``` nim
proc foo: Future[int] =
  code code code
  ...
  let v = waitFor bar()
  ...
  return v
```

I'd rather have an API that lets me experiment more easily with such changes in the compilation strategy.

Do you have any alternative proposal that can ensure all of the above? For the safety, I've given specific examples of problematic code that will be eliminated under the new APIs.

And finally, here is the full equivalent example under your alternative:

``` nim
var
  googlePageFut = httpRequest(&"http://google.com/?q={keyword}")
  jsonDataFut = jsonRpcCall("localhost/myApi")

await wait(googlePageFut, jsonDataFut)

let
  googlePage = googlePageFut.read 
  jsonData = jsonDataFut.read

return googlePage.status == 200 and not jsonData.hasKey("error")
```

Why shouldn't we strive to eliminate this boilerplate?

**Comment by cheatfate** [comment]
@zah, i think you missing not something but alot, or maybe somebody misled you with this optimizations. We already talked with @yglukhov, that if variable passes through state, it become part of closure environment, not stack, so all optimizations will not give you so many benefits as you expect.

Also i think you need to read more about `rust` implementation `tokio`, because i dont want to style like in rust:
```nim
  read(x).after(read(y).after(read(z)))`
```

**Comment by zah** [comment]
@cheatfate, I know precisely how the compiler works, but you are not paying enough attention to what I'm saying:

> At the moment this is still the heap-allocated environment of the closure iterator, but this can be attacked with further optimizations in the compiler (escape analysis for the closure objects, better support for closure/iterator monomorphism in the run-time code)

**Comment by cheatfate** [comment]
@zah you know precisely how the compiler works, and i know precisely how async works.
The style of Rust API is consequence of using `zero-cost abstractions`, almost equal approach is made for javascript's `Promise`.
I dont like such approach, because it pretty hard to understand and use.

**Comment by dom96** [comment]
I haven't read all of the discussion so forgive me if I'm missing something, but I already have a few things I would like to note about the whole reasoning for this issue: "Current Problems".

> Consider the following simple async proc:
>
> ```nim
>proc terminateConnection(s: AsyncSocket) {.async.} =
>  var myDisconnectMsg: array[X, byte]
>  prepareDisconnect(myDisconnectMsg)
>  var res = s.send(addr MyDisconnectMsg[], X) # oops, forgot to call await here 
>  s.close()
>```
>It showcases two critical problems trigged by a simple omission of await in the code:
>
> *  The socket will be closed prematurely.
> *  The send operation will be working with bogus data.

I am against any mixing of pointers and async. If you're doing that then you are on your own and should bear the brunt of the consequences.

Separately I do recognise the problem and it similar to the problem of users not knowing better and using ``discard`` on a future (instead of ``asyncCheck``). But I think the solution to this is straightforward: destructors. Once a future is out of scope a callback can be set on it to ensure an exception is raised and that it does not get lost. Of course, I'm waiting for destructors to be stable to implement that.

Once that works, the above code should fail with an exception.

I'll read the rest of this as soon as I can as I am open to alternative solutions. But from what skimming this has seemingly revealed, this is suggesting really big changes to fix this one relatively minor issue.

**Comment by dom96** [comment]
So I read the article as well and I must say that the "nursery" idea proposed can be easily implemented on top of async await. I don't see a reason to restrict async await when these nurseries can be implemented (and very easily too from what I can tell).

**Comment by dom96** [comment]
Why would (3) be a possibility? Does Windows just ignore a closed socket and report that it sent a message even though it didn't?

**Comment by zah** [comment]
In an offline discussion, I promised to explain how an async accept loop may work under the new scheme. It was pointed out that if the code is written in the style of the old APIs, enforced awaits will lead to accepting and processing only one connection at a time:

``` nim
while true:
  var conn = await socket.accept()
  await conn.process() # while this is awaited, we are not accepting new connections
```

The solve this problem, you must use the new async support enabled in `parallel` blocks. In general, the parallel blocks provide semantics that very closely ressemble the style of programming used in the old async APIs with `spawn` being roughly similar to `asyncCheck`. When you stumble on a problem, you should always ask yourself if the escape hatch provided by parallel blocks in enough to solve it. Here is how the solution here looks like:

``` nim
parallel:
  while true:
    var conn = await socket.accept()
    spawn conn.process() # this gets added to the list of outstanding async ops
                         # of the block and the processing continues immediately
```

The parallel block won't be exited until all outstanding operations are complete. In this sense, it can be argued that the new API provides an additional safety benefit over the ad-hoc processing used in the old async APIs. With the old APIs, a shutdown procedure won't do anything about the outstanding async tasks by default, while here we clearly decide whether we want to wait for all the operations to finish or whether we wan to cancel them by using `break` to exit the loop.

Please also note that all the optimization benefits of the new scheme are still realized, because any local variable declared before the parallel or within the parallel block is guaranteed to outlive all the processing of the async operations started in the block and this allows us to pass pointers to such variables to the said async operations (enabling efficient `openarray` and `var` parameters and allowing us to store the futures and their results on the stack).

**Comment by iffy** [comment]
Ignorant, fly-by question: how much of this could be implemented as a package without change to stdlib? I'm interested in using a system like the Python Trio system, because I *think* (but don't know) that it would simplify some of my async code.  Anything I could do to help?

**Comment by zah** [comment]
A certain version of it can be implemented without any changes to the compiler. Depending on exactly how much integration with parallel blocks is desired (in other words whether you want to mix async I/O with computational tasks backed by a thread pool), some compiler support can improve the system (and it's more about enforcing safety than about expressiveness).
